RFC for Waiting for Individual Tasks in task_group (#1862)
uxlfoundation/oneTBB: rfcs/proposed/task_group_wait_single_task/README.md

* 6 [Alternative Implementation Approaches](#alternative-implementation-approaches)
* 6.1 [Assign Single ``wait_context`` to the Task](#assign-single-wait_context-to-the-task)
* 6.2 [Avoid ``wait_context`` Usage](#avoid-wait_context-usage)

## Introduction
The ``tbb::task_group`` class represents a collection of tasks that execute concurrently. Tasks can be dynamically added to the group during execution.

The [Task Group Dynamic Dependencies preview feature](../task_group_dynamic_dependencies/README.md) extends the ``task_group`` API to support task dependencies, ensuring that a *successor* task begins execution only after all its *predecessors* have completed.

The current ``task_group`` API allows waiting for the completion of all tasks in the group using ``task_group::wait`` or ``task_group::run_and_wait``.
The waiting thread may participate in executing other tasks within the ``task_arena``, which may not be related to the specific ``task_group``.

Although waiting for all tasks suits many use cases, some workflows would benefit from the ability to wait for individual tasks.

Consider an example using an out-of-order SYCL queue.

```cpp

> sycl::queue q{sycl::property::queue::in_order{false}};

Reviewer: I do not think the `in_order` property can be constructed as `false`; see https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#api:property-queue-in-order.

// Creating multiple tasks for the queue
sycl::event task1 = q.single_task(task1_body);
sycl::event task2 = q.memset(some_memory_ptr, 0, 1024);

sycl::event task3 = q.submit([&](sycl::handler& h) {
    h.depends_on(std::vector<sycl::event>{task1, task2});

    // task1_body executes
    // some_memory_ptr is set before task3_body runs
    h.single_task(task3_body);
});

> task3.wait(); // wait for event 

Reviewer: ```suggestion
task3.wait(); // wait a single task
```

// Submit additional work
q.wait();
```

Consider implementing a similar model using ``tbb::task_group``. The queue can be represented as a ``task_group`` object, where tasks correspond to command groups in the queue, and events are modeled using ``tbb::task_completion_handle``s.

```cpp
using event = tbb::task_completion_handle;

class queue {
    template <typename Body>
    event single_task(Body body) {
        tbb::task_handle task = m_task_group.defer(body);
        tbb::task_completion_handle comp_handle = task;
        m_task_group.run(std::move(task));

>         return {comp_handle};

Reviewer: ```suggestion
        return comp_handle;
```

    }

    void wait() { m_task_group.wait(); }
private:
    tbb::task_group m_task_group;
};
```

The ``sycl::handler::depends_on`` functionality can be implemented using the ``tbb::task_group::set_task_order`` API:

```cpp
class handler {
    void depends_on(const std::vector<event>& events) {
        for (event& e : events) {
            tbb::task_group::set_task_order(e, m_root);
        }
    }

    tbb::task_handle m_root;
};
```

Since the example above also uses ``sycl::event::wait`` to wait for the completion of a single task (``task3``), implementing this behavior with
a ``task_group`` requires support for waiting on a single ``task_completion_handle`` object.

## Proposal

The proposal aims to extend ``tbb::task_group`` and ``tbb::task_arena`` APIs by introducing new waiting functions that accept a
``task_completion_handle`` argument.
These functions exit when the task represented by the handle argument is completed or when the group execution is cancelled.

> The completion of other tasks within the ``task_group`` is not guaranteed.

Reviewer 2: ```suggestion
Other tasks within the ``task_group`` may not be complete when these new waiting function return.
```

Since the new waiting functions track the progress of a single task, returning a ``task_group_status`` may be misleading.
If the group execution is cancelled, the tracked task may still execute, and returning ``canceled`` does not accurately reflect the task's 
completion status.
If execution is not cancelled, the function would need to track whether other tasks remain in the group and return ``not_complete`` if any are still pending.

If the task group execution is canceled and the originally executed task has transferred its completion to another task that was canceled, the
waiting function will return ``task_status::canceled``.

To address this, a new enum ``task_status`` is proposed to track the status of the awaited task. ``task_status::complete`` indicates that the tracked 
task was executed and completed, while ``task_status::canceled`` signifies that the task was not executed due to group cancellation.

The proposed API is summarized below:

```cpp
// Defined in <oneapi/tbb/task_group.h>
namespace oneapi {
namespace tbb {

enum class task_status {
    not_complete,
    complete,
    canceled
};

class task_group {

>     task_status wait_task(task_completion_handle& comp_handle);

Reviewer 2: I can't remember if we previously discussed this but `wait_task` seems to me like a getter that returns the wait task. Shouldn't this be something like `wait_for_task`.

>     task_status run_and_wait_task(task_handle&& handle);

Author: Done

Returns: ``task_status::completed`` if the task was executed and ``task_status::canceled`` otherwise.

Introducing ``wait_for`` in the ``this_task_arena`` namespace is unnecessary, as it would be functionally identical to calling the ``task_group::wait_for`` API directly.

## Future Enhancements

### Run-and-wait methods for ``task_arena``

It may be beneficial to introduce APIs that simplify task submission into ``task_arena`` and support waiting for task completion.

This functionality would be semantically similar to:

```cpp
tbb::task_group tg;
tbb::task_arena arena;
tbb::task_handle task = tg.defer(...);

tbb::task_completion_handle comp_handle = task;
arena.enqueue(std::move(task));
task_status status = arena.wait_for(comp_handle);
```

One possible name for such an extension is ``task_arena::enqueue_and_wait_for(task_handle&& task)``.
This function is semantically close to ``task_arena::execute`` (which may submit the task and wait for its completion), but it guarantees that
the task is enqueued into the arena.

A similar function could also be added to the ``tbb::this_task_arena`` namespace.

### Work Isolation Issue

> Similar to other ``task_arena::wait_for`` overload that take a ``task_group`` argument, it makes sense to also consider work isolation for

Reviewer: ```suggestion
Similar to another ``task_arena::wait_for`` overload that take a ``task_group`` argument, it makes sense to also consider work isolation for
```

the APIs introduced in this proposal.
If the wait is unrestricted, the waiting thread may execute any task from the arena, including those unrelated to the completion of the awaited
task or its associated ``task_group``.

Ideally, the waiting thread should only execute tasks that contribute to the completion of the awaited task. In such a model, all tasks
within the same subgraph would need to share a common isolation tag. In theory, a successor could inherit the isolation tag from

> its predecessor. However, multiple predecessors will have different tags.

Reviewer: ```suggestion
its predecessor. However, multiple predecessors may have different tags.
```

When the calling thread invokes ``run_and_wait_task`` call, the ``begin_task`` is executed. After executing its body,
it notifies its single successor node in the list. The ``middle_task``'s dependency count becomes zero, and the task is bypassed.

When ``middle_task`` completes, it notifies its list, which contains both a successor node and a waiter node. The successor node notifies
``end_task``, making it available for execution. The waiter node decrements the reference count in the associated ``wait_context``, allowing the
corresponding ``run_and_wait_task`` call to complete.
The notified ``wait_context`` is the same context awaited by the calling thread.

If ``end_task`` is bypassed, the calling thread will not exit until all tasks in the graph have completed. If additional tasks follow ``end_task``,
the thread will only exit after each is bypassed sequentially.

However, since the current thread notifies the ``wait_context`` it is waiting on via the corresponding waiter node, it is more appropriate to avoid
bypassing the next task. Instead, the task should be spawned, and ``run_and_wait_task`` should exit after executing ``middle_task``.

> For the initial implementation, it is proposed to completely avoid bypassing the task returned from the notification list and to spawn it.

Reviewer: ```suggestion
For the initial implementation, it is proposed to completely avoid bypassing the task returned from the notification list and to spawn it instead.
```

To address this and to make new waiting functions consistent in the return type with other functions in ``task_group`` and ``task_arena``,
it is proposed to introduce a new status ``task_complete`` that indicates that the awaited task is completed.
It is unknown either the ``task_group`` is cancelled or not, since the task can still be completed even if the group execution was cancelled.

If the task group execution is canceled and the originally executed task has transferred its completion to another task that was canceled, the
waiting function will return ``task_group_status::canceled``.

The proposed API is summarized below:

```cpp
// Defined in <oneapi/tbb/task_group.h>
namespace oneapi {
namespace tbb {

enum task_group_status { // Existing API
    not_complete,        // Existing API
    complete,            // Existing API
    canceled,            // Existing API
    task_complete        // Proposed API 
}

class task_group {
    task_group_status wait_for_task(task_completion_handle& comp_handle);
    task_group_status run_and_wait_for_task(task_handle&& handle);
};

// Defined in <oneapi/tbb/task_arena.h>
class task_arena {

>     task_group_status wait_for(task_completion_handle& comp_handle);

Reviewer 2: Ok so, the `task_arena::wait_for` that receives a task completion handle is an overload of `task_arena::wait_for`  that receives a task group. So a consistent name there. Do we want consistency with names in `task_group`, which now has `task_group::wait_for_task` and `task_group::wait`. Even though I recently suggested `task_group::wait_for_task`, would it be better as `task_group::wait_for`?  And then `task_group::run_and_wait_for`?

The completion of other tasks within the ``task_group`` is not guaranteed.

To simplify the submission and waiting process for a single task, it is also proposed to introduce submit-and-wait functions that accept a
``task_handle`` argument.

If task completion is transferred to another task using ``tbb::task_group::transfer_completion_to(other_task)``, the waiting functions will wait for
the completion of ``other_task``.

The current waiting functions return a ``task_group_status``, which can have three possible values:
* ``tbb::task_group_status::not_complete`` - The group execution is not cancelled, and not all tasks in the group have completed.
* ``tbb::task_group_status::complete`` - The group execution is not cancelled, and all tasks in the group have completed.
* ``tbb::task_group_status::canceled`` - The group execution is cancelled, and the completion status of tasks is unknown.

Since the new waiting functions track the progress of a single task, returning a ``task_group_status`` may be misleading.
If the group execution is cancelled, the tracked task may still execute, and returning ``canceled`` does not accurately reflect the task's 
completion status.

> If execution is not cancelled, the function would need to track whether other tasks remain in the group and return ``not_complete` if any are still pending.

Reviewer 2: Will transfer of a task_completion lead to any confusion around cancellation, if for example, the original task has executed but the task that was transfered to is canceled?  The user will see canceled as a status even though the initial task did complete. I think it's fine and will just need to be well documented in transfer.

Similar to other ``task_arena::wait_for`` overload that take a ``task_group`` argument, it makes sense to also consider work isolation for
the APIs introduced in this proposal.
If the wait is unrestricted, the waiting thread may execute any task from the arena, including those unrelated to the completion of the awaited
task or its associated ``task_group``.

Ideally, the waiting thread should only execute tasks that contribute to the completion of the awaited task. In such a model, all tasks
within the same subgraph would need to share a common isolation tag. In theory, a successor could inherit the isolation tag from
its predecessor. However, multiple predecessors will have different tags.

This would require changes to the current isolation mechanism to support multiple isolation tags, allowing a thread waiting for a successor
to execute tasks with any of its predecessor's tags. 

As an initial step, it makes sense to isolate the waiting thread so that it only executes tasks related to the same ``task_group`` as the awaited task - similar to the improvement described in the [RFC for another overload of ``task_arena::wait_for``](../task_arena_waiting/task_group_interop.md).

> ## Exit Criteria and Open Questions

Reviewer 2: Is the intention to go through an experimental phase, or straight to production?

            return {nullptr, true};
        }
    }
};

struct wait_list_node : notification_list_node {
    wait_context my_wait_context{1};

    notify_result notify(bool is_cancel_signal) override {
        my_wait_context.release();
        if (r1::get_current_wait_context() == &my_wait_context) {
            // Signal the notification step to stop bypass loop
            return {nullptr, false};
        } else {
            return {nullptr, true};
        }
    }
};

// called from task::execute or task::cancel
task_handle_task* notify_list(notification_list_node* node, bool is_cancel_signal) {
    task_handle_task* bypass_task = nullptr;
    bool bypass_allowed = true;

    while (node != nullptr) {
        notification_list_node::notify_result result = node->notify(is_cancel_signal);

        if (!result.second) bypass_allowed = false;

        if (bypass_task == nullptr) {
            bypass_task = result.first;
        } else if (result.first) {
            spawn(*result.first);
        }

        node = node->next();
    }

    if (bypass_task && !bypass_allowed) {
        spawn(*bypass_task);
        return nullptr;
    } else {
        return bypass_task;
    }
}
```

> This solution does not restrict the thread that is notified by another thread of the task completion from entering the bypass loop of the currently

Reviewer 2: This line is hard to follow.

Author: Rephrased and added more details.

* ``tbb::task_group_status::not_complete`` - The group execution is not cancelled, and not all tasks in the group have completed.
* ``tbb::task_group_status::complete`` - The group execution is not cancelled, and all tasks in the group have completed.
* ``tbb::task_group_status::canceled`` - The group execution is cancelled, and the completion status of tasks is unknown.

Since the new waiting functions track the progress of a single task, returning a ``task_group_status`` may be misleading.
If the group execution is cancelled, the tracked task may still execute, and returning ``canceled`` does not accurately reflect the task's 
completion status.
If execution is not cancelled, the function would need to track whether other tasks remain in the group and return ``not_complete` if any are still pending.

If the task group execution is canceled and the originally executed task has transferred its completion to another task that was canceled, the
waiting function will return ``task_status::canceled``.

To address this, a new enum ``task_status`` is proposed to track the status of the awaited task. ``task_status::complete`` indicates that the tracked 

> task was executed and completed, while ``task_status::canceled`` signifies that the task was not executed due to group cancellation.

Reviewer: There can be a semantic difference between a task being executed and a task being complete, if the completion was transferred. For me, "executed" means just "the task body has been run" while "complete" means logical completion, or essentially "task waiters and successors receive a signal". Since `wait_for` has the latter semantics, I really prefer "complete" or better "task_complete". It is also consistent with `transfer_this_task_completion_to`.

> For `complete`, I find it a bit misleading to have a status of task_group that knows nothing about the actual status of the group and serves the status of the task instead.

This is why I suggest to name it `task_group_status::task_complete` to indicate this is about a task and not the whole group.
And I still see no reason to have a separate enum class while an extra enum value seems to work just as good. Yet another reason for the single enum is the consistency of `task_arena::wait_for` overloads.

> task was executed and completed, while ``task_status::canceled`` signifies that the task was not executed due to group cancellation.

Reviewer 2: I understand your point, that `task_group_status::task_complete` can mean that the task completion handle was signaled because the task or the target of last transfer has finished executing. But how can a user know if the "completed" task has generated a full and valid result? It might have short-circuited due to a canceled task_group. I suppose we can say that `task_complete` only means that the task or transferred task signaled the completion handle and leave it up to the user to determine if that task was successful and finished its work. But I think it still would be better to provide a different name rather than using a form of `complete` which is used by `task_group::wait` to indicate that the work was finished successfully. It doesn't have to be `executed`.

The proposed API is summarized below:

```cpp
// Defined in <oneapi/tbb/task_group.h>
namespace oneapi {
namespace tbb {

enum class task_status {
    not_complete,
    complete,
    canceled
};

class task_group {
    task_status wait_for(task_completion_handle& comp_handle);

>     task_status run_and_wait_for(task_handle&& handle);

Reviewer 2: If it's a function, then there cannot be predecessors since no task_completion_handle was created. So you'd execute that function as a task and wait for just that task (which has no dependencies). So why run and wait for it in the task_group at all? Why not just call the function? But then maybe it makes sense if the completion is transferred to work created inside of the task?

>     task_status run_and_wait_for(task_handle&& handle);

Reviewer: I am fine to keep it as an open question for the future.

```cpp
task_status wait_for(task_completion_handle& comp_handle);
```

Waits for the completion of the task represented by ``comp_handle``.
If completion was transferred to another task using ``tbb::task_group::transfer_completion_to``, the function waits for the completion of that task.

Returns ``task_status::completed`` if the task was executed, or ``task_status::canceled`` if it was not.

```cpp
task_status run_and_wait_for(task_handle&& handle);
```

Schedules the task represented by ``handle`` for execution (if it has no unresolved dependencies), and waits for its completion.

This is semantically equivalent to: ``task_completion_handle comp_handle = handle; run(std::move(handle)); wait_for(comp_handle)``.

Returns ``task_status::completed`` if the task was executed, or ``task_status::canceled`` if it was not.

### Member functions of ``task_arena`` class

```cpp
task_status wait_for(task_completion_handle& comp_handle);
```

Waits for the completion of the task represented by ``comp_handle``.
If completion was transferred to another task using ``tbb::task_group::transfer_completion_to``, the function waits for completion of that task.

> This is semantically equivalent to: ``execute([&] { tg.wait_for(comp_handle); })``.

Author: With our current implementation, `task_group` is not required for a single-task waiting. All we need is a `task_dynamic_state` and the associated `task_group_context`, both can be obtained from the task object.

In theory, we can implement `task_group::wait_task` (and even `run_and_wait_task` if we omit the "same task group" check) as a static function. But I have proposed member functions for consistency with other waiting functions.

I am not sure if implementing `task_arena::wait_for` wihout the task group is possible for any other TBB implementation. From the perspective of the further inclusion into oneTBB specification, it may make sense to add `task_group` argument into this function and keep it unused in our implementation. What do you think?

    task_handle_task* bypass_task = nullptr;
    bool bypass_allowed = true;

    while (node != nullptr) {
        notification_list_node::notify_result result = node->notify(is_cancel_signal);

        if (!result.second) bypass_allowed = false;

        if (bypass_task == nullptr) {
            bypass_task = result.first;
        } else if (result.first) {
            spawn(*result.first);
        }

        node = node->next();
    }

    if (bypass_task && !bypass_allowed) {
        spawn(*bypass_task);
        return nullptr;
    } else {
        return bypass_task;
    }
}
```

While the proposed solution avoids task bypassing when the waiting thread executes the task it is explicitly waiting for, bypassing remains
unrestricted in other scenarios. For example, if the waiting thread executes a task unrelated to the awaited one, such as a ``parallel_for``
task in the same arena, it may still enter the bypass loop, even if the reference count of the awaited ``wait_context`` is zero.

An alternative approach to address this limitation is to implement a general mechanism within the scheduler that forces the thread to exit the

> bypass loop and spawn the returned task if further execution should not be continued (i.e., ``waiter.continue_execution()`` returns ``false``).

Author: I would try and prefer this approach in the long run as well. The only side-effect of this approach is potential behavior change in general.
E.g. a thread that is waiting for `parallel_for` to complete is taking a task belonging to other `parallel_for` and enters the bypass loop. This change will allow to avoid execution of the unrelated tasks and exit the first parallel for once it is really completed without the delay. 
Both solution with saving the innermost wait context and avoiding bypass in case of any waiters in the list would not affect the general behavior of TBB master and worker threads.

> ## Alternative Implementation Approaches

Reviewer: What are benefits and downsides of the alternative approaches to the recommended one?

Author: The main benefit of the recommended implementation approach is that waiting for completion is guaranteed to be implemented using a single `r1::wait` or `r1::run_and_wait` in case of transferring. 

For both alternative approaches, we will need to switch to another waiting in case of transferring:

```cpp
task_dynamic_state* state = comp_handle.get_dynamic_state();

r1::wait(state->get_wait_context());

while (state->was_transferred()) {
    state = state->get_new_completion_point();
    r1::wait(state->get_wait_context());
}
```

With the recommended approach, in case of transferring the wait context pointer is migrating between tasks, hence we don't need to double check if we completion was transferred. If the wait context was released, the completion is guaranteed to happen (no matter of which "final" task in the transfer chain).

Another benefit is that the wait context is created only when the wait was requested (that is not true for the first alternative approach). 

I will add more details on the benefits and downsides into the RFC.

The ``get_status_of`` function can also be implemented as a member function of the ``task_completion_handle`` class.

An alternative is having two separate functions: ``is_completed`` and ``is_canceled``. The advantage of this approach is that ``task_group_status`` does not need to be used. However,
the example above would require two function calls:

```cpp
task_completion_handle ch = task;
if (!is_completed(ch) && !is_canceled(ch)) {
    tg.wait_for_task(ch);
}
```

Both ``is_completed`` and ``is_canceled`` can also be implemented as member functions of the ``task_completion_handle`` class.

### Feature Test Macro

It is proposed to add an explicit feature-test macro to determine the presence of the feature:

```cpp
#define TBB_PREVIEW_TASK_GROUP_EXTENSIONS 1
#include <oneapi/tbb/task_group.h>

#if TBB_TASK_GROUP_WAIT_INDIVIDUAL_TASKS >= 202xxx // Some documented value
group.wait_for_task(comp_handle);
#else
// Usage of other APIs or self-written workarounds
#endif
```

An alternative is to use ``TBB_VERSION`` macro for this purpose:

```cpp
#if TBB_VERSION >= VER // VER is the first TBB version where waiting for individual tasks was released
```

### Summary

```cpp

> // Defined in <oneapi/tbb/task_group.h>

Author: Fixed

> #define TBB_TASK_GROUP_WAIT_FOR_INDIVIDUAL_TASKS 202xxx

Reviewer 2: I agree that they are different. I think both features could have separate stages of enabling. So it might be easier to track that (as a user) if they are separate. For example, if new run and wait methods are added. So two things like TBB_HAS_TASK_DEPENDENCIES and TBB_HAS_WAIT_FOR_INDIVIDUAL_TASKS seems cleaner to me.

> #define TBB_TASK_GROUP_WAIT_FOR_INDIVIDUAL_TASKS 202xxx

Reviewer: FWIW, I have no strong preference for this very name.
